datarobot_batch_scoring/network.py

```diff
@@ -218,66 +286,110 @@ def __iter__(self):
+                                      ''.format(self.network_status.value))
+                        self.network_status.value = b"R"
+                        f = self._executor.submit(self._request, r)
+                        f.add_done_callback(self.request_cb)
+                        self.futures.append(f)
+                        break
+                    else:
+                        self.ui.debug('state: {} -> F'
+                                      ''.format(self.network_status.value))
+                        self.network_status.value = b"F"
+                        wait(self.futures, return_when=FIRST_COMPLETED)
+                yield
         #  wait for all batches to finish before returning
+        self.network_status.value = b"W"
         while self.futures:
             f_len = len(self.futures)
             self.futures = [i for i in self.futures if not i.done()]
             if f_len != len(self.futures):
-                self._ui.debug('Waiting for final requests to finish. '
-                               'remaining requests: {}'
-                               ''.format(len(self.futures)))
+                self.ui.debug('Waiting for final requests to finish. '
+                              'remaining requests: {}'
+                              ''.format(len(self.futures)))
             wait(self.futures, return_when=FIRST_COMPLETED)
+        self.network_status.value = b"D"
         yield True
 
     def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        self._executor.shutdown(wait=True)
+        if self.proc and self.proc.is_alive():
+            self.proc.terminate()
+
+    def run(self, dry_run=False):
+        if dry_run:
+            i = 0
+            for _ in self.perform_requests(True):
+                i += 1
+
+            return i
+
+        self._executor = ThreadPoolExecutor(self.concurrency)
+        self.session = requests.Session()
+        adapter = requests.adapters.HTTPAdapter(
+            pool_connections=self.concurrency, pool_maxsize=self.concurrency)
+        self.session.mount('http://', adapter)
+        self.session.mount('https://', adapter)
+
+        t0 = time()
+        i = 0
+        r = None
+        for r in self.perform_requests():
+            i += 1
+            self.ui.info('{} responses sent | time elapsed {}s'
+                         .format(i, time() - t0))
+
+        if r is True:
+            i -= 1
+
+        self.progress_queue.put((ProgressQueueMsg.NETWORK_DONE, {
+            "ret": r,
+            "processed": self.n_requests,
+            "retried": self.n_retried,
+            "consumed": self.n_consumed,
+            "rusage": get_rusage(),
+        }))
+
+    def go(self, dry_run=False):
+        if dry_run:
+            return self.run(dry_run)
+
+        self.ui.set_next_UI_name('network')
+        self.proc = \
+            multiprocessing.Process(target=self.run,
```

Comment:
From Kiev office scoring on ORM instance with compression gets 130 vs 160 seconds.
So I'm not so sure it does not give much to us..